File: prosessering/src/main/java/no/nav/familie/prosessering/internal/TaskWorker.java
package no.nav.familie.prosessering.internal;

import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.Metrics;
import no.nav.familie.prosessering.AsyncTask;
import no.nav.familie.prosessering.TaskBeskrivelse;
import no.nav.familie.prosessering.TaskFeil;
import no.nav.familie.prosessering.domene.Status;
import no.nav.familie.prosessering.domene.Task;
import no.nav.familie.prosessering.domene.TaskRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;
import org.springframework.aop.framework.AopProxyUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.annotation.AnnotationUtils;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import static no.nav.familie.log.mdc.MDCConstants.MDC_CALL_ID;

@Service
class TaskWorker {

    private static final MdcExtendedLogContext LOG_CONTEXT = MdcExtendedLogContext.getContext("prosess");
    private static final Logger log = LoggerFactory.getLogger(TaskWorker.class);
    private static final Logger secureLog = LoggerFactory.getLogger("secureLogger");
    private final TaskRepository taskRepository;
    private Map<String, AsyncTask> tasktypeMap = new HashMap<>();
    private Map<String, Integer> maxAntallFeilMap = new HashMap<>();
    private final HashMap<String, Counter> feiledeTasks = new HashMap<>();

    @Autowired
    public TaskWorker(TaskRepository taskRepository, List<AsyncTask> taskTyper) {
        this.taskRepository = taskRepository;
        taskTyper.forEach(this::kategoriserTask);
    }

    private void kategoriserTask(AsyncTask task) {
        final Class<?> aClass = AopProxyUtils.ultimateTargetClass(task);
        final var annotation = AnnotationUtils.findAnnotation(aClass, TaskBeskrivelse.class);
        Objects.requireNonNull(annotation, "annotasjon mangler");
        tasktypeMap.put(annotation.taskType(), task);
        maxAntallFeilMap.put(annotation.taskType(), annotation.maxAntallFeil());

        feiledeTasks.put(annotation.taskType(), Metrics.counter("mottak.kontantstotte.feilede.tasks", "status", annotation.taskType(), "beskrivelse", annotation.beskrivelse()));
    }

    @Async("taskExecutor")
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void doTask(Long henvendelseId) {
        Objects.requireNonNull(henvendelseId, "id kan ikke være null");
        doActualWork(henvendelseId);
    }

    // For testing
    void doActualWork(Long henvendelseId) {
        final var startTidspunkt = System.currentTimeMillis();
        Integer maxAntallFeil = 0;
        var taskDetails = taskRepository.findById(henvendelseId).orElseThrow();

        initLogContext(taskDetails);
        try {

            secureLog.trace("Behandler task='{}'", taskDetails);
            taskDetails.behandler();
            taskDetails = taskRepository.saveAndFlush(taskDetails);

            // finn tasktype
            AsyncTask task = finnTask(taskDetails.getType());
            maxAntallFeil = finnMaxAntallFeil(taskDetails.getType());

            // execute
            task.preCondition(taskDetails);
            task.doTask(taskDetails);
            task.postCondition(taskDetails);
            task.onCompletion(taskDetails);

            taskDetails.ferdigstill();
            secureLog.trace("Ferdigstiller task='{}'", taskDetails);
            taskRepository.saveAndFlush(taskDetails);
            taskRepository.flush();
            secureLog.info("Fullført kjøring av task '{}', kjøretid={} ms", taskDetails, (System.currentTimeMillis() - startTidspunkt));
        } catch (Exception e) {
            taskDetails.feilet(new TaskFeil(taskDetails, e), maxAntallFeil);
            // lager metrikker på tasks som har feilet max antall ganger.
            if (taskDetails.getStatus() == Status.FEILET) {
                feiledeTasks.get(taskDetails.getType()).increment();
            }
            secureLog.warn("Fullført kjøring av task '{}', kjøretid={} ms, feilmelding='{}'", taskDetails, (System.currentTimeMillis() - startTidspunkt), e);
            taskRepository.save(taskDetails);
        } finally {
            clearLogContext();
        }
    }

    private void clearLogContext() {
        LOG_CONTEXT.clear();
        MDC.remove(MDC_CALL_ID);
    }

    private void initLogContext(Task taskDetails) {
        MDC.put(MDC_CALL_ID, taskDetails.getCallId());
        LOG_CONTEXT.add("task", taskDetails.getType());
    }

    private AsyncTask finnTask(String taskType) {
        if (!tasktypeMap.containsKey(taskType)) {
            throw new IllegalArgumentException("Ukjent tasktype " + taskType);
        }
        return tasktypeMap.get(taskType);
    }

    private Integer finnMaxAntallFeil(String taskType) {
        if (!maxAntallFeilMap.containsKey(taskType)) {
            throw new IllegalArgumentException("Ukjent tasktype " + taskType);
        }
        return maxAntallFeilMap.get(taskType);
    }
}

File: prosessering/src/main/java/no/nav/familie/prosessering/internal/TaskWorker.kt
package no.nav.familie.prosessering.internal

import io.micrometer.core.instrument.Counter
import io.micrometer.core.instrument.Metrics
import no.nav.familie.prosessering.AsyncTaskStep
import no.nav.familie.prosessering.TaskFeil
import no.nav.familie.prosessering.TaskStepBeskrivelse
import no.nav.familie.prosessering.domene.Status
import no.nav.familie.prosessering.domene.Task
import no.nav.familie.prosessering.domene.TaskRepository
import org.slf4j.LoggerFactory
import org.springframework.aop.framework.AopProxyUtils
import org.springframework.core.annotation.AnnotationUtils
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional

@Service
class TaskWorker(private val taskRepository: TaskRepository, taskStepTyper: List<AsyncTaskStep>) {

    private val taskStepMap: Map<String, AsyncTaskStep>

    private val maxAntallFeilMap: Map<String, Int>
    private val triggerTidVedFeilMap: Map<String, Long>
    private val feiltellereForTaskSteps: Map<String, Counter>
    private val fullførttellereForTaskSteps: Map<String, Counter>
    private val feiletStatusMap: Map<String, Status>

    init {
        val tasksTilTaskStepBeskrivelse: Map<AsyncTaskStep, TaskStepBeskrivelse> = taskStepTyper.associateWith { task ->
            val aClass = AopProxyUtils.ultimateTargetClass(task)
            val annotation = AnnotationUtils.findAnnotation(aClass, TaskStepBeskrivelse::class.java)
            requireNotNull(annotation) { "annotasjon mangler" }
            annotation
        }
        taskStepMap = tasksTilTaskStepBeskrivelse.entries.associate { it.value.taskStepType to it.key }
        maxAntallFeilMap = tasksTilTaskStepBeskrivelse.values.associate { it.taskStepType to it.maxAntallFeil }
        feiletStatusMap = tasksTilTaskStepBeskrivelse.values.associate { it.taskStepType to it.feiletStatus }
        triggerTidVedFeilMap = tasksTilTaskStepBeskrivelse.values.associate { it.taskStepType to it.triggerTidVedFeilISekunder }
        feiltellereForTaskSteps = tasksTilTaskStepBeskrivelse.values.associate {
            it.taskStepType to Metrics.counter("mottak.feilede.tasks",
                                               "status",
                                               it.taskStepType,
                                               "beskrivelse",
                                               it.beskrivelse)
        }
        fullførttellereForTaskSteps = tasksTilTaskStepBeskrivelse.values.associate {
            it.taskStepType to Metrics.counter("prosesering.fullført.tasks",
                                               "status",
                                               it.taskStepType,
                                               "beskrivelse",
                                               it.beskrivelse)
        }
    }

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    fun doActualWork(taskId: Long) {

        val taskOptional = taskRepository.findById(taskId)

        if (taskOptional.isEmpty) {
            return
        }

        val task = taskOptional.get()
        if (task.status != Status.PLUKKET) {
            return // en annen pod har startet behandling
        }

        task.behandler()

        // finn tasktype
        val taskStep = finnTaskStep(task.taskStepType)

        // execute
        taskStep.preCondition(task)
        taskStep.doTask(task)
        taskStep.postCondition(task)
        taskStep.onCompletion(task)

        task.ferdigstill()
        secureLog.trace("Ferdigstiller task='{}'", task)

        finnFullførtteller(task.taskStepType).increment()
    }

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    fun doFeilhåndtering(task: Task, e: Exception) {
        val maxAntallFeil = finnMaxAntallFeil(task.taskStepType)
        val feiletStatus = finnFeiletStatus(task.taskStepType)
        secureLog.trace("Behandler task='{}'", task)

        if (task.status == Status.MANUELL_OPPFØLGING) {
            // I dette tilfellet skal det kun logges, det skal ikke sjekkes mot maxAntallFeil eller oppdatere status.
            task.feiletMedStatusManuellOppfølging(TaskFeil(task, e))
        } else {
            task.feilet(TaskFeil(task, e), maxAntallFeil, feiletStatus)
            // lager metrikker på tasks som har feilet max antall ganger.
            if (task.status == feiletStatus) {
                finnFeilteller(task.taskStepType).increment()
                log.error("Task ${task.id} av type ${task.taskStepType} har feilet og satt til status $feiletStatus. Sjekk familie-prosessering for detaljer")
            }
        }
        task.triggerTid = task.triggerTid?.plusSeconds(finnTriggerTidVedFeil(task.taskStepType))
        taskRepository.save(task)
        secureLog.info("Feilhåndtering lagret ok {}", task)
    }

    private fun finnTriggerTidVedFeil(taskType: String): Long {
        return triggerTidVedFeilMap[taskType] ?: 0
    }

    private fun finnFeilteller(taskType: String): Counter {
        return feiltellereForTaskSteps[taskType] ?: error("Ukjent tasktype $taskType")
    }

    private fun finnFullførtteller(taskType: String): Counter {
        return fullførttellereForTaskSteps[taskType] ?: error("Ukjent tasktype $taskType")
    }

    private fun finnMaxAntallFeil(taskType: String): Int {
        return maxAntallFeilMap[taskType] ?: error("Ukjent tasktype $taskType")
    }

    private fun finnFeiletStatus(taskType: String): Status {
        return feiletStatusMap[taskType] ?: error("Ukjent tasktype $taskType")
    }

    private fun finnTaskStep(taskType: String): AsyncTaskStep {
        return taskStepMap[taskType] ?: error("Ukjent tasktype $taskType")
    }

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    fun markerPlukket(id: Long) {
        val taskOptional = taskRepository.findById(id)

        if (taskOptional.isEmpty) {
            return
        }

        val task = taskOptional.get()
        if (task.status.kanPlukkes()) {
            task.plukker()
        }
    }

    companion object {
        private val secureLog = LoggerFactory.getLogger("secureLogger")
        private val log = LoggerFactory.getLogger(this::class.java)
    }
}
